Enable the Array API even without omnisci, and fix/detect memory leaks

antocuni opened the pull request:
The PR is ready to review.
This is related to #378, although probably not a complete fix.

### Summary of changes

  - add `rbc.stdlib.array_api`, which rbc users can import to use the array-like API 
  - the array api is no longer specific to omnisci, now it works also in normal `@rjit` functions. This makes it much easier to develop and test
  - introduce `rbclib`. This is the embryo of a "runtime library" to be called by the code generated by rbc. In particular, it provides `rbclib_allocate_varlen_buffer`, which can be used to allocate arrays for non-omnisci targets
  - store the names of the allocation/deallocation array functions in the `TargetInfo`. This makes it possible to call the appropriate `allocate_varlen_buffer` function depending on whether we are targeting omnisci or not
  - introduce the "debug allocator" which keeps track of allocations/deallocations and detect memory leaks (see e.g. #378 )

A note about `rbclib`. I implemented it with CFFI, which is by far the easiest way to compile custom C code and load it inside the Python process. The choice is explained in detail in a docstring:
https://github.com/antocuni/rbc/blob/2b47ac863a81e1d0465466489c4d4d7f4f4309f1/rbc/rbclib/__init__.py#L1-L40

Moreover, thanks to CFFI we can write `extern "C+Python"` functions, which can be implemented in Python but called from C/llvm as if they were written in C. This is how I implemented the debug allocator (which is much easier to implement in Python than in C).

### Things to discuss

  - `rbc.stdlib`: I think that the current layout of `rbc` is a bit messy because it mixes internal stuff of the compiler and "external libraries" which can be imported and used by the final users. My proposal is that everything which can be imported by the users should live in `rbc.stdlib`, but this is up to discussion.


### Follow up

Currently, `xp.Array` still leaks. Next PRs should finally find a way to avoid leaking memory

guilhermeleobas commented on rbc/remotejit.py:
```diff
@@ -404,7 +411,7 @@ def bar(a, b):
 
     def __init__(self, host='localhost', port=11532,
-                 local=False, debug=False):
+                 local=False, debug=False, use_debug_allocator=False):
```
What about use `debug=True` to enable the debug allocator?

antocuni replied:
I think it's cleaner to have a separate flag for the debug allocator, especially because it works only with `local=True`: what should it do with `debug=True, local=False`? Complain or ignore the debug allocator? Better to be explicit, IMHO.

Also it seems that `debug` is a weird flag in `rbc`. E.g. it seems to me that it only increases verbosity but doesn't actually affect the generated code.

pearu replied:
"use_debug_allocator" contains two verbs "use" and "debug". Suggest using "debug_allocator".

antocuni replied:
uhm, this is more an english detail than a technical one: in this sentence, "debug" is an adjective, like when we say e.g. "the debug version" or so. To it's meaning is "use the debug version of the allocator", at least as I intended it.
That say, I can change it if you want

pearu replied:
yes, this is English. According to google search, "debug" is most commonly a verb (can also be a noun).
On the other hand, I see what you mean: you have introduced `debug_allocator` module and intend to say to use this module. However, in this context, I think the existence of this module is irrelevant as an internal implementation detail and for users, the intent of the flag is "to debug the memory allocation/deallocation consistency".
If this makes sense, let's use "debug_allocator".

antocuni replied:
I started to rename `use_debug_allocator` into `debug_allocator` but I quickly encountered a stylistic problem:
```python
from .rbclib import debug_allocator
[...]
    @dispatchermethod
    def call(self, fullname: str, arguments: tuple) -> Data:
        [...]
        if self.debug_allocator:
            leak_detector = debug_allocator.new_leak_detector()
```

i.e., now the `debug_allocator` name is overloaded with two meaning: the module itself and a bool indicating whether to use it or now. I think the above code is confusing, because it looks similiar-but-not-quite to a standard python pattern, which IMHO is confusing:
```python
if self.debug_allocator:
    leak_detector = self.debug_allocator.new_leak_detector()
```

no, now I am -0.5 to rename it

antocuni replied:
> Rename the module?

good idea. I renamed it as `tracing_allocator`, which is a much better name. Thanks for convincing me to do this :)

antocuni commented:
@guilhermeleobas @pearu the PR is ready for review and hopefully merging.
I edited the first comment so that it now contains all the relevant updated information.

pearu reviewed (changes requested):
Great work, @antocuni! I have a number of comments from reading the code.

guilhermeleobas commented on rbc/rbclib/__init__.py:
```diff
@@ -0,0 +1,60 @@
+     done by load_inside_llvm(), which is called from
+     libfuncs.RBCLib.__init__().
+
+NOTE: rbclib is based on CFFI, but this is not strictly necessary: in theory,
+it is enough to dlopen() a C library which contains the desired
+functions. However, using CFFI has many advantages:
```
For RBC users who want to use it with OmniSciDB, the CFFI package is unnecessary, I guess. @antocuni, would it be possible to make it an optional dependency?

guilhermeleobas replied:
> Although one option would be to replace _rbclib.c functionality with a hand-written LLVM IR or use numba to create the corresponding functions

One could use the code on `ctools.py` to compile C code to LLVM IR.

pearu replied:
cffi cannot build extension modules in runtime, hence the suggestion cannot be implemented.

guilhermeleobas commented on rbc/rbclib/debug_allocator.py:
```diff
@@ -0,0 +1,100 @@
+from ._rbclib import lib, ffi
+
+
+class DebugAllocatorError(Exception):
+    pass
+
+
+class InvalidFreeError(DebugAllocatorError):
+    pass
+
+
+class MemoryLeakError(DebugAllocatorError):
+
+    def __init__(self, leaks):
+        lines = [f'Found {len(leaks)} memory leaks:']
+        for addr, seq in leaks:
+            lines.append(f'    {addr} (seq = {seq})')
+        message = '\n'.join(lines)
+        super().__init__(message)
+        self.leaks = leaks
```
Move to `errors.py`

antocuni replied:
e8b2593

antocuni commented on .github/workflows/rbc_test.yml:
```diff
@@ -34,11 +34,15 @@ jobs:
-        numba-version: [0.54, 0.53]
+        # os: [ubuntu-latest, windows-latest, macos-latest]
+        # python-version: [3.9, 3.8, 3.7]
+        # numba-version: [0.54, 0.53]
```
uh, good catch. Fixed by d349f68

antocuni commented on doc/conf.py:
```diff
@@ -15,7 +15,7 @@
 import sys
 import datetime
-sys.path.insert(0, os.path.abspath('../'))
+#sys.path.insert(0, os.path.abspath('../'))  # XXX
```
fixed by d8ad66a

antocuni commented on rbc/libfuncs.py:
```diff
@@ -307,3 +309,13 @@ def check(self, fname):
+    from rbc.rbclib import FUNCTION_NAMES as _function_names
+
+    def __init__(self):
+        import rbc.rbclib
```
a961da9

antocuni commented on rbc/libfuncs.py:
```diff
@@ -307,3 +309,13 @@ def check(self, fname):
+class RBCLib(Library):
+
+    name = 'rbclib'
+    from rbc.rbclib import FUNCTION_NAMES as _function_names
```
it's needed by `Library`. If you look, all other `Library` subclasses defines the same. I've added an explanatory comment in cbb33d1

antocuni commented on rbc/rbclib/__init__.py:
```diff
@@ -0,0 +1,60 @@
+
+  - _rbclib_build.py is the CFFI build script which is called by setup.py
+
+  - _rbclib.abi3.so is the native extension produced by _rbclib_build.py and that
```
84a499f

antocuni commented on rbc/rbclib/debug_allocator.py:
```diff
@@ -0,0 +1,100 @@
+        self.alive_memory = {}  # {address: seq}
+
+    def record_allocate(self, addr):
+        self.seq += 1
```
good idea, added by 2e0595f

antocuni commented on rbc/remotejit.py:
```diff
@@ -7,11 +7,17 @@
 from .utils import get_local_ip
 from .targetinfo import TargetInfo
+from .rbclib import debug_allocator
+# XXX WIP: the OmnisciCompilerPipeline is no longer omnisci-specific because
+# we support Arrays even without omnisci, so it must be renamed and moved
+# somewhere elsef
```
if you agree, I'll do the proper rename/move in the next PR where I'll also need to fix it

pearu replied:
Yes, a follow-up PR sounds good to me.

pearu commented on rbc/rbclib/_rbclib.c:
```diff
@@ -0,0 +1,28 @@
+    // XXX: currently there is a bug in the FreeOmnisciBuffer and the free()
+    // is never called.  We should remove the abort() as soon as this code is
+    // actually called
+    abort();
```
I am looking forward to the next PR :)
IIUC, udf/udtf developers need to explicitly call free on arrays that are not returned.

antocuni replied:
we need to discuss&decide what is the best strategy to solve the issue:
1. we require them to call free explicitly
2. we try to automagically determine which arrays needs to be freed

I think that at a first glance, 1 is preferable

antocuni commented on rbc/rbclib/intrinsic.py:
```diff
@@ -0,0 +1,23 @@
+@extending.intrinsic
+def add_ints(typingctx, a_type, b_type):
+    if (a_type, b_type) != (nb_types.int64, nb_types.int64):
+        return None
```
I did `return None` because the [numba docs](https://numba.pydata.org/numba-doc/latest/extending/high-level.html#implementing-intrinsics) contains the following:

> The return values of func should be a 2-tuple of expected type signature, and a code-generation function that will passed to lower_builtin(). For an unsupported operation, return None.

What is the different between `return None` and `raise NumbaTypeError()`? When should I use one instead of the other?

guilhermeleobas replied:
They both work. But in a type error, you explicitly say why the function didn't match. I guess for the tests, this is not needed since you call the function with the correct argument types.

antocuni replied:
thanks, fixed by 3b621d0

antocuni commented:
@pearu @guilhermeleobas thanks for the review.
I think I addressed most of your comments and replied to a couple which require further discussion so it's probably ready for another round of review

pearu reviewed (changes requested):
In this review round, I have tried to make cffi optional. The aim is to keep rbc-feedstock to be noarch.

antocuni commented on setup.py:
```diff
@@ -71,6 +71,7 @@ def setup_package():
         ],
         packages=find_packages(),
         package_data={'': ['*.thrift']},
+        cffi_modules=['rbc/rbclib//_rbclib_build.py:ffibuilder'],
```
note that this is not strictly needed. `cffi_modules` is a setuptools keyword and it's used only if CFFI is installed at build time. If not, it's just ignored

pearu commented:
Rename the module?

On Fri, Dec 17, 2021, 12:04 PM Antonio Cuni ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In rbc/remotejit.py
> <https://github.com/xnd-project/rbc/pull/380#discussion_r771260318>:
>
> > @@ -404,7 +411,7 @@ def bar(a, b):
>      typesystem_aliases = dict()
>
>      def __init__(self, host='localhost', port=11532,
> -                 local=False, debug=False):
> +                 local=False, debug=False, use_debug_allocator=False):
>
> I started to rename use_debug_allocator into debug_allocator but I
> quickly encountered a stylistic problem:
>
> from .rbclib import debug_allocator
> [...]
>     @dispatchermethod
>     def call(self, fullname: str, arguments: tuple) -> Data:
>         [...]
>         if self.debug_allocator:
>             leak_detector = debug_allocator.new_leak_detector()
>
> i.e., now the debug_allocator name is overloaded with two meaning: the
> module itself and a bool indicating whether to use it or now. I think the
> above code is confusing, because it looks similiar-but-not-quite to a
> standard python pattern, which IMHO is confusing:
>
> if self.debug_allocator:
>     leak_detector = self.debug_allocator.new_leak_detector()
>
> no, now I am -0.5 to rename it
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/xnd-project/rbc/pull/380#discussion_r771260318>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADCF3AZR3BBLNM3B3U2PI3URMDLBANCNFSM5IMXKW3A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

guilhermeleobas commented:
Failure doesn't seem to be related to this PR. I've run the tests locally using the software versions of the failing container and cannot reproduce the error

<details><summary>Tests report</summary>
```
$ pytest rbc/tests/test_omnisci_column_basic.py -rs -sv
============================================================================================== test session starts ==============================================================================================
platform linux -- Python 3.8.12, pytest-6.2.5, py-1.11.0, pluggy-1.0.0 -- /home/guilhermeleobas/miniconda3/envs/rbc/bin/python
cachedir: .pytest_cache
rootdir: /home/guilhermeleobas/git/rbc, configfile: setup.cfg
collecting ...  OmnisciDB version (5, 6, 4, '-20210823-58f0d7365c')
collected 95 items

rbc/tests/test_omnisci_column_basic.py::test_sizer_row_multiplier_orig PASSED
rbc/tests/test_omnisci_column_basic.py::test_sizer_row_multiplier_param1 PASSED
rbc/tests/test_omnisci_column_basic.py::test_sizer_row_multiplier_param2 PASSED
rbc/tests/test_omnisci_column_basic.py::test_sizer_constant_parameter PASSED
rbc/tests/test_omnisci_column_basic.py::test_sizer_return_size PASSED
rbc/tests/test_omnisci_column_basic.py::test_rowmul_add_columns PASSED
rbc/tests/test_omnisci_column_basic.py::test_rowmul_return_mixed_columns PASSED
rbc/tests/test_omnisci_column_basic.py::test_rowmul_return_multi_columns[1-1--1] PASSED
rbc/tests/test_omnisci_column_basic.py::test_rowmul_return_multi_columns[1-1-3] PASSED
rbc/tests/test_omnisci_column_basic.py::test_rowmul_return_multi_columns[1-2--1] PASSED
rbc/tests/test_omnisci_column_basic.py::test_rowmul_return_multi_columns[1-2-3] PASSED
rbc/tests/test_omnisci_column_basic.py::test_rowmul_return_multi_columns[2-1--1] PASSED
rbc/tests/test_omnisci_column_basic.py::test_rowmul_return_multi_columns[2-1-3] PASSED
rbc/tests/test_omnisci_column_basic.py::test_rowmul_return_multi_columns[2-2--1] PASSED
rbc/tests/test_omnisci_column_basic.py::test_rowmul_return_multi_columns[2-2-3] PASSED
rbc/tests/test_omnisci_column_basic.py::test_rowmul_return_multi_columns[3-1--1] PASSED
rbc/tests/test_omnisci_column_basic.py::test_rowmul_return_multi_columns[3-1-3] PASSED
rbc/tests/test_omnisci_column_basic.py::test_rowmul_return_multi_columns[3-2--1] PASSED
rbc/tests/test_omnisci_column_basic.py::test_rowmul_return_multi_columns[3-2-3] PASSED
rbc/tests/test_omnisci_column_basic.py::test_rowmul_return_multi_columns[4-1--1] PASSED
rbc/tests/test_omnisci_column_basic.py::test_rowmul_return_multi_columns[4-1-3] PASSED
rbc/tests/test_omnisci_column_basic.py::test_rowmul_return_multi_columns[4-2--1] PASSED
rbc/tests/test_omnisci_column_basic.py::test_rowmul_return_multi_columns[4-2-3] PASSED
rbc/tests/test_omnisci_column_basic.py::test_issue173[1] PASSED
rbc/tests/test_omnisci_column_basic.py::test_issue173[2] PASSED
rbc/tests/test_omnisci_column_basic.py::test_issue173[3] PASSED
rbc/tests/test_omnisci_column_basic.py::test_redefine PASSED
rbc/tests/test_omnisci_column_basic.py::test_overload_nonuniform[1] XFAIL (Test failing due to the introduction of default sizer. See PR 313)
rbc/tests/test_omnisci_column_basic.py::test_overload_nonuniform[2] XFAIL (Test failing due to the introduction of default sizer. See PR 313)
rbc/tests/test_omnisci_column_basic.py::test_overload_nonuniform[3] XFAIL (Test failing due to the introduction of default sizer. See PR 313)
rbc/tests/test_omnisci_column_basic.py::test_overload_uniform PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[avg-] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[avg-groupby] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[min-] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[min-groupby] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[max-] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[max-groupby] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[sum-] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[sum-groupby] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[count-] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[count-groupby] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[approx_count_distinct-] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[approx_count_distinct-groupby] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[sample-] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[sample-groupby] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[single_value-] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[single_value-groupby] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[stddev-] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[stddev-groupby] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[stddev_pop-] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[stddev_pop-groupby] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[stddev_samp-] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[stddev_samp-groupby] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[variance-] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[variance-groupby] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[var_pop-] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[var_pop-groupby] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[var_samp-] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[var_samp-groupby] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[correlation-] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[correlation-groupby] SKIPPED (correlation-groupby test crashes CPU-only omnisci server [rbc issue 237])
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[corr-] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[corr-groupby] SKIPPED (corr-groupby test crashes CPU-only omnisci server [rbc issue 237])
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[covar_pop-] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[covar_pop-groupby] SKIPPED (covar_pop-groupby test crashes CPU-only omnisci server [rbc issue 237])
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[covar_samp-] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_aggregate[covar_samp-groupby] SKIPPED (covar_samp-groupby test crashes CPU-only omnisci server [rbc issue 237])
rbc/tests/test_omnisci_column_basic.py::test_column_function[sin] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_function[cos] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_function[tan] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_function[asin] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_function[acos] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_function[atan] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_function[abs] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_function[ceil] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_function[degrees] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_function[exp] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_function[floor] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_function[ln] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_function[log] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_function[log10] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_function[radians] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_function[round] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_function[sign] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_function[sqrt] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_function[atan2] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_function[power] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_binary_operation[+] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_binary_operation[-] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_binary_operation[*] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_binary_operation[/] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_unary_operation[+] PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_unary_operation[-] PASSED
rbc/tests/test_omnisci_column_basic.py::test_create_as PASSED
rbc/tests/test_omnisci_column_basic.py::test_column_different_sizes PASSED

=============================================================================================== warnings summary ================================================================================================
rbc/tests/test_omnisci_column_basic.py::test_sizer_row_multiplier_orig
  /home/guilhermeleobas/git/rbc/rbc/omniscidb.py:778: UserWarning:
    thrift type map: update TExtArgumentType.ColumnListBool to 44 (was 42)
    thrift type map: update TExtArgumentType.ColumnListDouble to 43 (was 41)
    thrift type map: update TExtArgumentType.ColumnListFloat to 42 (was 40)
    thrift type map: update TExtArgumentType.ColumnListInt16 to 39 (was 37)
    thrift type map: update TExtArgumentType.ColumnListInt32 to 40 (was 38)
    thrift type map: update TExtArgumentType.ColumnListInt64 to 41 (was 39)
    thrift type map: update TExtArgumentType.ColumnListInt8 to 38 (was 36)
    thrift type map: add new member TExtArgumentType.TextEncodingDict16
    thrift type map: add new member TExtArgumentType.TextEncodingDict32
    thrift type map: add new member TExtArgumentType.TextEncodingDict8
    warnings.warn('\n  '.join([''] + messages))

rbc/tests/test_omnisci_column_basic.py::test_sizer_row_multiplier_orig
  /home/guilhermeleobas/git/rbc/rbc/omniscidb.py:719: UserWarning: thrift.TExtArgumentType.TextEncodingDict16(=36) value not in ext_arguments_map
    warnings.warn('thrift.TExtArgumentType.%s(=%s) value not '

rbc/tests/test_omnisci_column_basic.py::test_sizer_row_multiplier_orig
  /home/guilhermeleobas/git/rbc/rbc/omniscidb.py:719: UserWarning: thrift.TExtArgumentType.TextEncodingDict32(=37) value not in ext_arguments_map
    warnings.warn('thrift.TExtArgumentType.%s(=%s) value not '

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============================================================================================ short test summary info ============================================================================================
SKIPPED [1] rbc/tests/test_omnisci_column_basic.py:616: correlation-groupby test crashes CPU-only omnisci server [rbc issue 237]
SKIPPED [1] rbc/tests/test_omnisci_column_basic.py:616: corr-groupby test crashes CPU-only omnisci server [rbc issue 237]
SKIPPED [1] rbc/tests/test_omnisci_column_basic.py:616: covar_pop-groupby test crashes CPU-only omnisci server [rbc issue 237]
SKIPPED [1] rbc/tests/test_omnisci_column_basic.py:616: covar_samp-groupby test crashes CPU-only omnisci server [rbc issue 237]
============================================================================= 88 passed, 4 skipped, 3 xfailed, 3 warnings in 27.13s =============================================================================
```
</details>

pearu commented on rbc/remotejit.py:
```diff
@@ -419,11 +426,17 @@ def __init__(self, host='localhost', port=11532,
             host = get_local_ip()
 
+        if use_tracing_allocator and not local:
+            raise ValueError('use_tracing_allocator=True can be used only with local=True')
```
What restricts using the tracing allocator with a remote client?

antocuni replied:
it's not a fundamental restriction, but for now it's just a missing functionality. Currently it works this way:
1. the `TracingAllocator` records the allocations on the remote side (where the code is exectued)
2. the `LeakDetector` checks whether the `TracingAllocator` leaked anything inside a `with:` statement around a call site, which happens on the local side

To make it working with `local=False`, we would need to implement one or more thrift calls so that the `TracingAllocator` and the `LeakDetector` can communicate. I even started to do it, but then I realized that it was becoming unnecessarily complex and that for the sake of running our own tests `local=True` works well enough and it's much simpler to implement.

pearu replied:
Thanks! I created https://github.com/xnd-project/rbc/issues/391 to keep track of this limitation. The issue is not high-priority but a good-first-task.

pearu reviewed (approved):
LGTM! Thanks, @antocuni !
